<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Setup" data-toc-modified-id="Setup-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Setup---Debug" data-toc-modified-id="Setup---Debug-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Setup - Debug</a></span></li><li><span><a href="#Setup---Imports" data-toc-modified-id="Setup---Imports-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Setup - Imports</a></span></li><li><span><a href="#Setup---working-folder-paths" data-toc-modified-id="Setup---working-folder-paths-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Setup - working folder paths</a></span></li><li><span><a href="#Setup---logging" data-toc-modified-id="Setup---logging-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Setup - logging</a></span></li><li><span><a href="#Setup---virtualenv-jupyter-kernel" data-toc-modified-id="Setup---virtualenv-jupyter-kernel-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Setup - virtualenv jupyter kernel</a></span></li><li><span><a href="#Setup---Initialize-Django" data-toc-modified-id="Setup---Initialize-Django-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Setup - Initialize Django</a></span></li><li><span><a href="#Setup---Initialize-LoggingHelper" data-toc-modified-id="Setup---Initialize-LoggingHelper-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Setup - Initialize LoggingHelper</a></span></li><li><span><a href="#Setup---initialize-ProquestHNPNewspaper" data-toc-modified-id="Setup---initialize-ProquestHNPNewspaper-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Setup - initialize ProquestHNPNewspaper</a></span><ul class="toc-item"><li><span><a href="#load-from-database" data-toc-modified-id="load-from-database-2.8.1"><span class="toc-item-num">2.8.1&nbsp;&nbsp;</span>load from database</a></span></li><li><span><a href="#set-up-manually" data-toc-modified-id="set-up-manually-2.8.2"><span class="toc-item-num">2.8.2&nbsp;&nbsp;</span>set up manually</a></span></li></ul></li></ul></li><li><span><a href="#Find-articles-to-be-loaded" data-toc-modified-id="Find-articles-to-be-loaded-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Find articles to be loaded</a></span><ul class="toc-item"><li><span><a href="#Uncompress-files" data-toc-modified-id="Uncompress-files-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Uncompress files</a></span></li><li><span><a href="#Work-with-uncompressed-files" data-toc-modified-id="Work-with-uncompressed-files-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Work with uncompressed files</a></span></li><li><span><a href="#parse-and-load-XML-files" data-toc-modified-id="parse-and-load-XML-files-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>parse and load XML files</a></span></li><li><span><a href="#build-list-of-all-ObjectTypes" data-toc-modified-id="build-list-of-all-ObjectTypes-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>build list of all ObjectTypes</a></span><ul class="toc-item"><li><span><a href="#Build-sorted-list-of-object-types" data-toc-modified-id="Build-sorted-list-of-object-types-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Build sorted list of object types</a></span></li></ul></li><li><span><a href="#map-files-to-types" data-toc-modified-id="map-files-to-types-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>map files to types</a></span></li><li><span><a href="#map-publication-date-ranges-to-folders" data-toc-modified-id="map-publication-date-ranges-to-folders-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>map publication date ranges to folders</a></span></li></ul></li><li><span><a href="#TODO" data-toc-modified-id="TODO-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>TODO</a></span></li></ul></div>

# Introduction

- Back to [Table of Contents](#Table-of-Contents)

This is a notebook that expands on the OpenCalais code in the file `article_coding.py`, also in this folder.  It includes more sections on selecting publications you want to submit to OpenCalais as an example.  It is intended to be copied and re-used.

# Setup

- Back to [Table of Contents](#Table-of-Contents)

## Setup - Debug

- Back to [Table of Contents](#Table-of-Contents)

In [1]:
debug_flag = False

## Setup - Imports

- Back to [Table of Contents](#Table-of-Contents)

In [2]:
import datetime
import glob
import logging
import lxml
import os
import six
import xml
import xmltodict
import zipfile

## Setup - working folder paths

- Back to [Table of Contents](#Table-of-Contents)

What data are we looking at?

In [3]:
# paper identifier
paper_identifier = "BostonGlobe"
archive_identifier = "BG_20171002210239_00001"

# source
source_paper_folder = "/mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data"
source_paper_path = "{}/{}".format( source_paper_folder, paper_identifier )

# uncompressed
uncompressed_paper_folder = "/mnt/hgfs/projects/phd/proquest_hnp/uncompressed"
uncompressed_paper_path = "{}/{}".format( uncompressed_paper_folder, paper_identifier )

# make sure an identifier is set before you make a path here.
if ( ( archive_identifier is not None ) and ( archive_identifier != "" ) ):
    
    # identifier is set.
    source_archive_file = "{}.zip".format( archive_identifier )
    source_archive_path = "{}/{}".format( source_paper_path, source_archive_file )
    uncompressed_archive_path = "{}/{}".format( uncompressed_paper_path, archive_identifier )

#-- END check to see if archive_identifier present. --#

In [4]:
%pwd

'/home/jonathanmorgan/work/django/research/work/phd_work/data/article_loading/proquest_hnp/BostonGlobe'

In [5]:
# current working folder
current_working_folder = "/home/jonathanmorgan/work/django/research/work/phd_work/data/article_loading/proquest_hnp/{}".format( paper_identifier )
current_datetime = datetime.datetime.now()
current_date_string = current_datetime.strftime( "%Y-%m-%d" )

## Setup - logging

- Back to [Table of Contents](#Table-of-Contents)

configure logging for this notebook's kernel (If you do not run this cell, you'll get the django application's logging configuration.

In [6]:
logging_file_name = "{}/research-data_load-{}-{}.log.txt".format( current_working_folder, paper_identifier, current_date_string )
logging.basicConfig(
    level = logging.DEBUG,
    format = '%(asctime)s - %(levelname)s - %(name)s - %(message)s',
    filename = logging_file_name,
    filemode = 'w' # set to 'a' if you want to append, rather than overwrite each time.
)

## Setup - virtualenv jupyter kernel

- Back to [Table of Contents](#Table-of-Contents)

If you are using a virtualenv, make sure that you:

- have installed your virtualenv as a kernel.
- choose the kernel for your virtualenv as the kernel for your notebook (Kernel --> Change kernel).

Since I use a virtualenv, need to get that activated somehow inside this notebook.  One option is to run `../dev/wsgi.py` in this notebook, to configure the python environment manually as if you had activated the `sourcenet` virtualenv.  To do this, you'd make a code cell that contains:

    %run ../dev/wsgi.py
    
This is sketchy, however, because of the changes it makes to your Python environment within the context of whatever your current kernel is.  I'd worry about collisions with the actual Python 3 kernel.  Better, one can install their virtualenv as a separate kernel.  Steps:

- activate your virtualenv:

        workon research

- in your virtualenv, install the package `ipykernel`.

        pip install ipykernel

- use the ipykernel python program to install the current environment as a kernel:

        python -m ipykernel install --user --name <env_name> --display-name "<display_name>"
        
    `sourcenet` example:
    
        python -m ipykernel install --user --name sourcenet --display-name "research (Python 3)"
        
More details: [http://ipython.readthedocs.io/en/stable/install/kernel_install.html](http://ipython.readthedocs.io/en/stable/install/kernel_install.html)

## Setup - Initialize Django

- Back to [Table of Contents](#Table-of-Contents)

First, initialize my dev django project, so I can run code in this notebook that references my django models and can talk to the database using my project's settings.

In [7]:
# init django
django_init_folder = "/home/jonathanmorgan/work/django/research/work/phd_work"
django_init_path = "django_init.py"
if( ( django_init_folder is not None ) and ( django_init_folder != "" ) ):
    
    # add folder to front of path.
    django_init_path = "{}/{}".format( django_init_folder, django_init_path )
    
#-- END check to see if django_init folder. --#

In [8]:
%run $django_init_path

django initialized at 2019-08-15 05:15:59.419616


In [9]:
# context_text imports
from context_text.article_coding.article_coding import ArticleCoder
from context_text.article_coding.article_coding import ArticleCoding
from context_text.article_coding.open_calais_v2.open_calais_v2_article_coder import OpenCalaisV2ArticleCoder
from context_text.collectors.newsbank.newspapers.GRPB import GRPB
from context_text.collectors.newsbank.newspapers.DTNB import DTNB
from context_text.models import Article
from context_text.models import Article_Subject
from context_text.models import Newspaper
from context_text.shared.context_text_base import ContextTextBase

# context_text_proquest_hnp
from context_text_proquest_hnp.proquest_hnp_newspaper_helper import ProquestHNPNewspaperHelper

## Setup - Initialize LoggingHelper

- Back to [Table of Contents](#Table-of-Contents)

Create a LoggingHelper instance to use to log debug and also print at the same time.

Preconditions: Must be run after Django is initialized, since `python_utilities` is in the django path.

In [10]:
# python_utilities
from python_utilities.logging.logging_helper import LoggingHelper

# init
my_logging_helper = LoggingHelper()
my_logging_helper.set_logger_name( "proquest_hnp-article-loading-{}".format( paper_identifier ) )
log_message = None

## Setup - initialize ProquestHNPNewspaper

- Back to [Table of Contents](#Table-of-Contents)

Create an initialize an instance of ProquestHNPNewspaper for this paper.

### load from database

- Back to [Table of Contents](#Table-of-Contents)

In [11]:
my_paper = ProquestHNPNewspaperHelper()
paper_instance = my_paper.initialize_from_database( paper_identifier )
my_paper.source_all_papers_folder = source_paper_folder
my_paper.destination_all_papers_folder = uncompressed_paper_folder

In [12]:
print( my_paper )
print( paper_instance )

BostonGlobe from 1872 to 1985 - source path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe - dest path: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe
1 - BostonGlobe - Boston Globe, The


### set up manually

- Back to [Table of Contents](#Table-of-Contents)

In [11]:
my_paper = ProquestHNPNewspaperHelper()
my_paper.paper_identifier = paper_identifier
my_paper.source_all_papers_folder = source_paper_folder
my_paper.source_paper_path = source_paper_path
my_paper.destination_all_papers_folder = uncompressed_paper_folder
my_paper.destination_paper_path = uncompressed_paper_path
my_paper.paper_start_year = 1872
my_paper.paper_end_year = 1985

my_newspaper = Newspaper.objects.get( id = 6 )
my_paper.newspaper = my_newspaper

If desired, add to database.

In [12]:
phnp_newspaper_instance = my_paper.create_PHNP_newspaper()

In [13]:
print( phnp_newspaper_instance )

1 - BostonGlobe - Boston Globe, The


# Find articles to be loaded

- Back to [Table of Contents](#Table-of-Contents)

Specify which folder of XML files should be loaded into system, then process all files within the folder.

The compressed archives from proquest_hnp just contain publication XML files, no containing folder.

To process:

- **uncompresed paper folder ( `<paper_folder>` )** - make a folder in `/mnt/hgfs/projects/phd/proquest_hnp/uncompressed` for the paper whose data you are working with, named the same as the paper's folder in `/mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data`.

    - for example, for the Boston Globe, name it "`BostonGlobe`".

- **uncompressed archive folder ( `<archive_folder>` )** - inside a given paper's folder in uncompressed, for each archive file, create a folder named the same as the archive file, but with no ".zip" at the end.

    - For example, for the file "`BG_20171002210239_00001.zip`", make a folder named "`BG_20171002210239_00001`".
    - path should be "`<paper_folder>/<archive_name_no_zip>`.

- unzip the archive into this folder:

        unzip <path_to_zip> -d <archive_folder>



## Uncompress files

- Back to [Table of Contents](#Table-of-Contents)

See if the uncompressed paper folder exists.  If not, set flag and create it.

In [14]:
# create folder to hold the results of decompressing paper's zip files.
did_uncomp_paper_folder_exist = my_paper.make_dest_paper_folder()

EXISTS - Uncompressed paper folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe


For each *.zip file in the paper's source folder:

- parse file name from path returned by glob.
- parse the part before ".zip" from the file name.  This is referred to subsequently as the "archive identifier".
- check if folder named the same as the "archive identifier" is present.

    - If no:
    
        - create it.
        - then, uncompress the archive into it.
        
    - If yes:
    
        - output a message.  Don't want to uncompress if it was already uncompressed once.

In [15]:
# decompress the files
my_paper.uncompress_paper_zip_files()

==> zip file count: 474
----> processing file 1 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151210212722_00001.zip
==> file: BG_20151210212722_00001.zip
==> ID: BG_20151210212722_00001
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212722_00001
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212722_00001
------------------------------
----> processing file 2 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151210212824_00002.zip
==> file: BG_20151210212824_00002.zip
==> ID: BG_20151210212824_00002
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212824_00002
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212824_00002
------------------------------
----> processing file 3 of 474
==> path: /mnt/

==> file: BG_20151210223106_00006.zip
==> ID: BG_20151210223106_00006
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210223106_00006
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210223106_00006
------------------------------
----> processing file 71 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151210223207_00007.zip
==> file: BG_20151210223207_00007.zip
==> ID: BG_20151210223207_00007
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210223207_00007
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210223207_00007
------------------------------
----> processing file 72 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151210223208_00008.zip
==> file: BG_20151210223208_00008.zip
==> ID: BG_20151210223208_0000

==> file: BG_20151211000407_00010.zip
==> ID: BG_20151211000407_00010
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211000407_00010
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211000407_00010
------------------------------
----> processing file 147 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211001111_00001.zip
==> file: BG_20151211001111_00001.zip
==> ID: BG_20151211001111_00001
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211001111_00001
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211001111_00001
------------------------------
----> processing file 148 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211001213_00002.zip
==> file: BG_20151211001213_00002.zip
==> ID: BG_20151211001213_00

EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211020319_00010
------------------------------
----> processing file 229 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211020320_00011.zip
==> file: BG_20151211020320_00011.zip
==> ID: BG_20151211020320_00011
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211020320_00011
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211020320_00011
------------------------------
----> processing file 230 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211020421_00012.zip
==> file: BG_20151211020421_00012.zip
==> ID: BG_20151211020421_00012
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211020421_00012
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/proje

EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211032258_00012
------------------------------
----> processing file 297 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211032259_00013.zip
==> file: BG_20151211032259_00013.zip
==> ID: BG_20151211032259_00013
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211032259_00013
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211032259_00013
------------------------------
----> processing file 298 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211032301_00014.zip
==> file: BG_20151211032301_00014.zip
==> ID: BG_20151211032301_00014
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211032301_00014
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/proje

EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211042322_00013
------------------------------
----> processing file 360 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211042424_00014.zip
==> file: BG_20151211042424_00014.zip
==> ID: BG_20151211042424_00014
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211042424_00014
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211042424_00014
------------------------------
----> processing file 361 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211042525_00015.zip
==> file: BG_20151211042525_00015.zip
==> ID: BG_20151211042525_00015
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211042525_00015
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/proje

EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211052405_00018
------------------------------
----> processing file 440 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211052908_00001.zip
==> file: BG_20151211052908_00001.zip
==> ID: BG_20151211052908_00001
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211052908_00001
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211052908_00001
------------------------------
----> processing file 441 of 474
==> path: /mnt/hgfs/projects/phd/proquest_hnp/proquest_hnp/data/BostonGlobe/BG_20151211053011_00002.zip
==> file: BG_20151211053011_00002.zip
==> ID: BG_20151211053011_00002
==> TO: /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211053011_00002
EXISTS, so moving on - Uncompressed archive folder /mnt/hgfs/proje

## Work with uncompressed files

- Back to [Table of Contents](#Table-of-Contents)

Change working directories to the uncompressed paper path.

In [8]:
%cd $uncompressed_paper_path

/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe


In [9]:
%ls

BG_20151210212722_00001/  BG_20151211002934_00005/  BG_20151211034630_00005/
BG_20151210212824_00002/  BG_20151211003037_00006/  BG_20151211034731_00006/
BG_20151210212825_00003/  BG_20151211003138_00007/  BG_20151211034732_00007/
BG_20151210212926_00004/  BG_20151211003139_00008/  BG_20151211034833_00008/
BG_20151210212927_00005/  BG_20151211003240_00009/  BG_20151211034934_00009/
BG_20151210213028_00006/  BG_20151211004245_00001/  BG_20151211034935_00010/
BG_20151210213031_00007/  BG_20151211004347_00002/  BG_20151211035036_00011/
BG_20151210213132_00008/  BG_20151211004448_00003/  BG_20151211035037_00012/
BG_20151210213134_00009/  BG_20151211004549_00004/  BG_20151211035138_00013/
BG_20151210213235_00010/  BG_20151211004552_00005/  BG_20151211035239_00014/
BG_20151210213236_00011/  BG_20151211004653_00006/  BG_20151211035240_00015/
BG_20151210213337_00012/  BG_20151211004754_00007/  BG_20151211035341_00016/
BG_20151210213338_00013/  BG_20151211004755_00008/  BG_20151211035845_00001/

## parse and load XML files

- Back to [Table of Contents](#Table-of-Contents)

Load one of the files into memory and see what we can do with it.  Beautiful Soup?

Looks like the root element is "Record", then the high-level type of the article is "ObjectType".

ObjectType values:

- Advertisement
- ...

Good options for XML parser:

- `lxml.etree` - [https://stackoverflow.com/questions/12290091/reading-xml-file-and-fetching-its-attributes-value-in-python](https://stackoverflow.com/questions/12290091/reading-xml-file-and-fetching-its-attributes-value-in-python)
- `xmltodict` - [https://docs.python-guide.org/scenarios/xml/](https://docs.python-guide.org/scenarios/xml/)
- `beautifulsoup` using `lxml`

In [14]:
# loop over files in the current archive folder path.
object_type_to_count_map = my_paper.process_archive_object_types( uncompressed_archive_path )

Processing 5752 XML files in /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20171002210239_00001
----> XML file count: 5752
XML file count: 5752
Counters:
- Processed 5752 files
- No Record: 0
- No ObjectType: 0
- No ObjectType value: 0

ObjectType values and occurrence counts:
- A|d|v|e|r|t|i|s|e|m|e|n|t: 1902
- Article|Feature: 1792
- N|e|w|s: 53
- Commentary|Editorial: 36
- G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n: 488
- S|t|o|c|k| |Q|u|o|t|e: 185
- Advertisement|Classified Advertisement: 413
- E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c: 31
- Correspondence|Letter to the Editor: 119
- Front Matter|Table of Contents: 193
- O|b|i|t|u|a|r|y: 72
- F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y: 107
- I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h: 84
- Marriage Announcement|News: 6
- I|l|l|u|s|t|r|a|t|i|o|n: 91
- R|e|v|i|e|w: 133
- C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t: 30
- News|Legal Notice: 17


    Processing 5752 files in /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20171002210239_00001
    ----> XML file count: 5752

    Counters:
    - Processed 5752 files
    - No Record: 0
    - No ObjectType: 0
    - No ObjectType value: 0

    ObjectType values and occurrence counts:
    - A|d|v|e|r|t|i|s|e|m|e|n|t: 1902
    - Article|Feature: 1792
    - N|e|w|s: 53
    - Commentary|Editorial: 36
    - G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n: 488
    - S|t|o|c|k| |Q|u|o|t|e: 185
    - Advertisement|Classified Advertisement: 413
    - E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c: 31
    - Correspondence|Letter to the Editor: 119
    - Front Matter|Table of Contents: 193
    - O|b|i|t|u|a|r|y: 72
    - F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y: 107
    - I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h: 84
    - Marriage Announcement|News: 6
    - I|l|l|u|s|t|r|a|t|i|o|n: 91
    - R|e|v|i|e|w: 133
    - C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t: 30
    - News|Legal Notice: 17

## build list of all ObjectTypes

- Back to [Table of Contents](#Table-of-Contents)

Loop over all folders in the paper path.  For each folder, grab all files in the folder.  For each file, parse XML, then get the ObjectType value and if it isn't already in map of obect types to counts, add it.  Increment count.

From command line, in the uncompressed BostonGlobe folder:

    find . -type f -iname "*.xml" | wc -l

resulted in 11,374,500 articles.  That is quite a few.

In [14]:
xml_folder_list = glob.glob( "{}/*".format( uncompressed_paper_path ) )
print( "folder_list: {}".format( xml_folder_list ) )

folder_list: ['/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212722_00001', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212824_00002', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212825_00003', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212926_00004', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212927_00005', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213028_00006', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213031_00007', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213132_00008', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213134_00009', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213235_00010', '/mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210213236_00011', '/mnt/hgfs/project

In [ ]:
# build map of all object types for a paper to the overall counts of each
paper_object_type_to_count_map = my_paper.process_paper_object_types()

Processing 474 XML folders in /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe
==> Processing XML folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212722_00001 ( 1 of 474 ) @ 2019-08-15 05:19:18.880092
----> Processing complete @ 2019-08-15 05:20:43.997772 ( duration 0:01:25.117680 )

==> Processing XML folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212824_00002 ( 2 of 474 ) @ 2019-08-15 05:20:43.998197
----> Processing complete @ 2019-08-15 05:22:10.260852 ( duration 0:01:26.262655 )

==> Processing XML folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212825_00003 ( 3 of 474 ) @ 2019-08-15 05:22:10.261221
----> Processing complete @ 2019-08-15 05:24:01.944264 ( duration 0:01:51.683043 )

==> Processing XML folder /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151210212926_00004 ( 4 of 474 ) @ 2019-08-15 05:24:01.944842
----> Processing complete @ 2019-08-15 05:25:

XML file count: 5752
Counters:
- Processed 5752 files
- No Record: 0
- No ObjectType: 0
- No ObjectType value: 0

ObjectType values and occurrence counts:
- A|d|v|e|r|t|i|s|e|m|e|n|t: 2114224
- Feature|Article: 5271887
- I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h: 249942
- O|b|i|t|u|a|r|y: 625143
- G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n: 1083164
- S|t|o|c|k| |Q|u|o|t|e: 202776
- N|e|w|s: 140274
- I|l|l|u|s|t|r|a|t|i|o|n: 106925
- F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y: 386421
- E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c: 78993
- Editorial|Commentary: 156342
- C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t: 68356
- Classified Advertisement|Advertisement: 291533
- R|e|v|i|e|w: 86889
- Table of Contents|Front Matter: 69798
- Letter to the Editor|Correspondence: 202071
- News|Legal Notice: 24053
- News|Marriage Announcement: 41314
- B|i|r|t|h| |N|o|t|i|c|e: 926
- News|Military/War News: 3
- U|n|d|e|f|i|n|e|d: 5
- Article|Feature: 137526
- Front Matter|Table of Contents: 11195
- Commentary|Editorial: 3386
- Marriage Announcement|News: 683
- Correspondence|Letter to the Editor: 7479
- Legal Notice|News: 1029
- Advertisement|Classified Advertisement: 12163


### Build sorted list of object types

- Back to [Table of Contents](#Table-of-Contents)

In [1]:
# put the raw output from above in a list
raw_object_type_list = [ 'A|d|v|e|r|t|i|s|e|m|e|n|t: 2114224', 'Feature|Article: 5271887', 'I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h: 249942', 'O|b|i|t|u|a|r|y: 625143', 'G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n: 1083164', 'S|t|o|c|k| |Q|u|o|t|e: 202776', 'N|e|w|s: 140274', 'I|l|l|u|s|t|r|a|t|i|o|n: 106925', 'F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y: 386421', 'E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c: 78993', 'Editorial|Commentary: 156342', 'C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t: 68356', 'Classified Advertisement|Advertisement: 291533', 'R|e|v|i|e|w: 86889', 'Table of Contents|Front Matter: 69798', 'Letter to the Editor|Correspondence: 202071', 'News|Legal Notice: 24053', 'News|Marriage Announcement: 41314', 'B|i|r|t|h| |N|o|t|i|c|e: 926', 'News|Military/War News: 3', 'U|n|d|e|f|i|n|e|d: 5', 'Article|Feature: 137526', 'Front Matter|Table of Contents: 11195', 'Commentary|Editorial: 3386', 'Marriage Announcement|News: 683', 'Correspondence|Letter to the Editor: 7479', 'Legal Notice|News: 1029', 'Advertisement|Classified Advertisement: 12163' ]

In [25]:
# output variable
master_object_type_list = None

# declare variables
#raw_object_type_list = None
raw_object_type = None
object_type_part_list = None
object_type_to_count_map = None
object_type_value = None
object_type_count_string = None
object_type_count = None

# loop
master_object_type_list = []
object_type_to_count_map = {}
for raw_object_type in raw_object_type_list:
    
    # split on colon
    object_type_part_list = raw_object_type.split( ":" )
    
    # object type value - take the first thing, strip off spaces, and add it to list.
    object_type_value = object_type_part_list[ 0 ]
    object_type_value = object_type_value.strip()

    # object type value count - item 2 (index 1)
    object_type_count_string = object_type_part_list[ 1 ]
    object_type_count_string = object_type_count_string.strip()
    object_type_count = int( object_type_count_string )
    
    # add value to list.
    if ( object_type_value not in master_object_type_list ):
        
        # add it.
        master_object_type_list.append( object_type_value )
        
    else:
        
        # error.
        print( "ERROR - object type value {} in list more than once.  Hmmm.".format( object_type_value ) )
        
    #-- END check to see if value already in list. --#
    
    # add count to map.
    if ( object_type_value not in object_type_to_count_map ):
        
        # add count.
        object_type_to_count_map[ object_type_value ] = object_type_count
        
    else:
        
        # error.
        print( "ERROR - object type value {} already has count in map.  Hmmm.".format( object_type_value ) )
        
    #-- END check to see if value already in list. --#
    
#-- END loop over raw object types --#

# sort the list of object types
master_object_type_list.sort()

print( master_object_type_list )

['Advertisement|Classified Advertisement', 'Article|Feature', 'A|d|v|e|r|t|i|s|e|m|e|n|t', 'B|i|r|t|h| |N|o|t|i|c|e', 'Classified Advertisement|Advertisement', 'Commentary|Editorial', 'Correspondence|Letter to the Editor', 'C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t', 'Editorial|Commentary', 'E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c', 'Feature|Article', 'Front Matter|Table of Contents', 'F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y', 'G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n', 'I|l|l|u|s|t|r|a|t|i|o|n', 'I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h', 'Legal Notice|News', 'Letter to the Editor|Correspondence', 'Marriage Announcement|News', 'News|Legal Notice', 'News|Marriage Announcement', 'News|Military/War News', 'N|e|w|s', 'O|b|i|t|u|a|r|y', 'R|e|v|i|e|w', 'S|t|o|c|k| |Q|u|o|t|e', 'Table of Contents|Front Matter', 'U|n|d|e|f|i|n|e|d']


    ['Advertisement|Classified Advertisement', 'Article|Feature', 'A|d|v|e|r|t|i|s|e|m|e|n|t', 'B|i|r|t|h| |N|o|t|i|c|e', 'Classified Advertisement|Advertisement', 'Commentary|Editorial', 'Correspondence|Letter to the Editor', 'C|r|e|d|i|t|/|A|c|k|n|o|w|l|e|d|g|e|m|e|n|t', 'Editorial|Commentary', 'E|d|i|t|o|r|i|a|l| |C|a|r|t|o|o|n|/|C|o|m|i|c', 'Feature|Article', 'Front Matter|Table of Contents', 'F|r|o|n|t| |P|a|g|e|/|C|o|v|e|r| |S|t|o|r|y', 'G|e|n|e|r|a|l| |I|n|f|o|r|m|a|t|i|o|n', 'I|l|l|u|s|t|r|a|t|i|o|n', 'I|m|a|g|e|/|P|h|o|t|o|g|r|a|p|h', 'Legal Notice|News', 'Letter to the Editor|Correspondence', 'Marriage Announcement|News', 'News|Legal Notice', 'News|Marriage Announcement', 'News|Military/War News', 'N|e|w|s', 'O|b|i|t|u|a|r|y', 'R|e|v|i|e|w', 'S|t|o|c|k| |Q|u|o|t|e', 'Table of Contents|Front Matter', 'U|n|d|e|f|i|n|e|d']


## map files to types

- Back to [Table of Contents](#Table-of-Contents)

Choose a directory, then loop over the files in the directory to build a map of types to lists of file names.

In [15]:
# directory to work in.
uncompressed_archive_folder = "BG_20151211054235_00003"
uncompressed_archive_path = "{}/{}".format( uncompressed_paper_path, uncompressed_archive_folder )

In [16]:
# build map of file types to lists of files of that type in specified folder.
object_type_to_file_path_map = my_paper.map_archive_folder_files_to_types( uncompressed_archive_path )


In map_archive_folder_files_to_types:
XML file count: 25000
Counters:
- Processed 25000 files
- No Record: 0
- No ObjectType: 0
- No ObjectType value: 0

ObjectType values and occurrence counts:
- Feature|Article:
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501666496.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501667056.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501704736.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501705376.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501706296.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501707216.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501708776.xml
    - /mnt/hgfs/projects/phd/proquest_hnp/uncompre

In [29]:
# which types do we want to preview?
types_to_output = master_object_type_list
#types_to_output = [ 'Advertisement|Classified Advertisement' ]

# declare variables
xml_file_path_list = None
xml_file_path_example_list = None
xml_file_path = None
xml_file = None
xml_dict = None
xml_string = None

# loop over types
for object_type in types_to_output:
    
    # print type and count
    xml_file_path_list = object_type_to_file_path_map.get( object_type, [] )
    xml_file_path_example_list = xml_file_path_list[ : 10 ]
    print( "\n- {}:".format( object_type ) )
    for xml_file_path in xml_file_path_example_list:
        
        print( "----> {}".format( xml_file_path ) )

        # try to parse the file
        with open( xml_file_path ) as xml_file:

            # parse XML
            xml_dict = xmltodict.parse( xml_file.read() )
            
        #-- END with open( xml_file_path ) as xml_file: --#
            
        # pretty-print
        xml_string = xmltodict.unparse( xml_dict, pretty = True )

        # output
        print( xml_string )
        
    #-- END loop over example file paths. --#
    
#-- END loop over object types. --#


- Advertisement|Classified Advertisement:

- Article|Feature:

- A|d|v|e|r|t|i|s|e|m|e|n|t:
----> /mnt/hgfs/projects/phd/proquest_hnp/uncompressed/BostonGlobe/BG_20151211054235_00003/501691896.xml
<?xml version="1.0" encoding="utf-8"?>
<Record>
	<Version>Record_v1.0.xsd</Version>
	<RecordID>501691896</RecordID>
	<DateTimeStamp>20151211001353</DateTimeStamp>
	<ActionCode>change</ActionCode>
	<RecordTitle>Display Ad 8 -- No Title</RecordTitle>
	<Publication>
		<PublicationID>46044</PublicationID>
		<Title>Boston Daily Globe (1872-1922)</Title>
		<Qualifier>Boston, Mass.</Qualifier>
	</Publication>
	<Publisher>New York Times Company</Publisher>
	<AlphaPubDate>Apr 6, 1911</AlphaPubDate>
	<NumericPubDate>19110406</NumericPubDate>
	<SourceType>Historical Newspapers</SourceType>
	<ObjectType>Advertisement</ObjectType>
	<LanguageCode>ENG</LanguageCode>
	<StartPage>4</StartPage>
	<Pagination>4</Pagination>
	<URLDocView>http://search.proquest.com/docview/501691896/</URLDocView>
	<Products>
		<P

<?xml version="1.0" encoding="utf-8"?>
<Record>
	<Version>Record_v1.0.xsd</Version>
	<RecordID>502945091</RecordID>
	<DateTimeStamp>20151211001753</DateTimeStamp>
	<ActionCode>change</ActionCode>
	<RecordTitle>Editorial Cartoon 1 -- No Title</RecordTitle>
	<Publication>
		<PublicationID>46044</PublicationID>
		<Title>Boston Daily Globe (1872-1922)</Title>
		<Qualifier>Boston, Mass.</Qualifier>
	</Publication>
	<Publisher>New York Times Company</Publisher>
	<AlphaPubDate>May 4, 1915</AlphaPubDate>
	<NumericPubDate>19150504</NumericPubDate>
	<SourceType>Historical Newspapers</SourceType>
	<ObjectType>Editorial Cartoon/Comic</ObjectType>
	<LanguageCode>ENG</LanguageCode>
	<StartPage>7</StartPage>
	<Pagination>7</Pagination>
	<URLDocView>http://search.proquest.com/docview/502945091/</URLDocView>
	<Products>
		<Product>
			<ProductID>1006466</ProductID>
			<HasFullText>true</HasFullText>
		</Product>
		<Product>
			<ProductID>1007870</ProductID>
			<HasFullText>true</HasFullText>
		</Produc

<?xml version="1.0" encoding="utf-8"?>
<Record>
	<Version>Record_v1.0.xsd</Version>
	<RecordID>501968976</RecordID>
	<DateTimeStamp>20151211001512</DateTimeStamp>
	<ActionCode>change</ActionCode>
	<RecordTitle>NEWTON LEAGUE CHAMPIONS.</RecordTitle>
	<Publication>
		<PublicationID>46044</PublicationID>
		<Title>Boston Daily Globe (1872-1922)</Title>
		<Qualifier>Boston, Mass.</Qualifier>
	</Publication>
	<Publisher>New York Times Company</Publisher>
	<AlphaPubDate>Mar 28, 1912</AlphaPubDate>
	<NumericPubDate>19120328</NumericPubDate>
	<SourceType>Historical Newspapers</SourceType>
	<ObjectType>Image/Photograph</ObjectType>
	<LanguageCode>ENG</LanguageCode>
	<StartPage>7</StartPage>
	<Pagination>7</Pagination>
	<URLDocView>http://search.proquest.com/docview/501968976/</URLDocView>
	<Products>
		<Product>
			<ProductID>1006466</ProductID>
			<HasFullText>true</HasFullText>
		</Product>
		<Product>
			<ProductID>1007870</ProductID>
			<HasFullText>true</HasFullText>
		</Product>
		<Product

<?xml version="1.0" encoding="utf-8"?>
<Record>
	<Version>Record_v1.0.xsd</Version>
	<RecordID>501849696</RecordID>
	<DateTimeStamp>20151211001334</DateTimeStamp>
	<ActionCode>change</ActionCode>
	<RecordTitle>Obituary 7 -- No Title</RecordTitle>
	<Publication>
		<PublicationID>46044</PublicationID>
		<Title>Boston Daily Globe (1872-1922)</Title>
		<Qualifier>Boston, Mass.</Qualifier>
	</Publication>
	<Publisher>New York Times Company</Publisher>
	<AlphaPubDate>Dec 17, 1911</AlphaPubDate>
	<NumericPubDate>19111217</NumericPubDate>
	<SourceType>Historical Newspapers</SourceType>
	<ObjectType>Obituary</ObjectType>
	<LanguageCode>ENG</LanguageCode>
	<StartPage>13</StartPage>
	<Pagination>13</Pagination>
	<URLDocView>http://search.proquest.com/docview/501849696/</URLDocView>
	<Products>
		<Product>
			<ProductID>1006466</ProductID>
			<HasFullText>true</HasFullText>
		</Product>
		<Product>
			<ProductID>1007870</ProductID>
			<HasFullText>true</HasFullText>
		</Product>
		<Product>
			<Pr

## map publication date ranges to folders



# TODO

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- figure out which ObjectTypes to explore, pick a folder and just eyeball a few, to see what they look like.
- / map publication date ranges to files (for each folder, capture max and min pub date - element: "`NumericPubDate`").
- / make a little `proquest_hnp` django app to hold metadata about the publications.  Build on Context_Text:

    - // models:

        - papers in context_text.
        - object types for each paper.

            - pipe-delimited value.
            - display value
            - ?

        - folders for each paper.

            - folder path
            - min pub date
            - max pub_date

        - paths to files in each folder?
        - 
        
    - // code:
    
        - class for dealing with files and folders - migrate the code from jupyter notebooks here, instance variables for folder settings, logging, etc.
        
    - admin:
    
        - make custom admins for everything so I can control what is on the list page, make lookups for foreign keys, etc. (need to have counts on the 
        - add object type inlines for papers and archives so you can see related object types.
        - add paper inline to object types, so you can see which papers use each inline.
        - ...?
        
- // make the `summarize_archive_files` method also create Archive records in django if none exist.
- / run summarize on all Boston Globe articles, store the results, so we can start to look at article quality in different eras.